### DISCUSSION TITLE
How to use and update the UI output for as a backend variable

### DISCUSSION THREAD

**Discussion by yipeng-jiang** [zauberzeug/nicegui#594] [Q&A] [answered]
Hey there,

I came from streamlit world. In streamlit I can dirrectly use the value from a streamlit compoent for backend calculation and use the calculated value to update other frontend components. For example

```py
import streamlit as st

id = st.selectbox(label ='model', ['A' 'B', 'C'] ")

// Below is a sudo code for a funciton to query a rest endpoint, which has to be ran on the server side.
data = query_rest_endpoint(id)

st.table(data)
```
I know streamlit rerun the entire python script each time when a component value changes,  So not sure how to do it in Nicegui? 

Thanks

**Comment by falkoschindler** [comment, marked as answer]
Good question, @yipeng-jiang! I guess many developers coming from Streamlit might be confused at first, especially if they know Python primarily through the Streamlit context. Although their approach seams appealing at first, there are [several major problems](https://github.com/zauberzeug/nicegui/discussions/21#discussioncomment-2209467) that led us to build a different, more classical API.

In NiceGUI you declare event handlers that are called, e.g., if a value changes. Such event handlers can be lambda statements (like `on_change` in the example below) or normal functions (like `on_click`):

```py
s = ui.select(['A', 'B', 'C'], label='Choose an option:',
              on_change=lambda e: result.set_text(f'You selected: {e.value}')).classes('w-48')
result = ui.label()

def handle_click():
    ui.notify('The button was clicked!')
    s.set_value('B')

ui.button('Select "B"', on_click=handle_click)
```

**Reply by yipeng-jiang** [reply]
I see. In order to update a UI component, we need to exiplicitly change the UI component property like `grid.options` or a method to change its value like `s.set_value('B')` in the event handler function and then optionally call update() method to UI Components to update.  Thank you very much. It is definitely a smarter way to avoid re-runing the entire script.  

Just currious. How did you design the data communication between frontend and backend under the hood. Is it every variable declared in the backend automatically generating a fastapi route attached to it which a comparable frontend variable is also genereated and able to fecth the data from this api route.

**Reply by rodja** [reply]
We sync the state between frontend and backend via websocket. Via the `.update()` call on an element we send the whole state of the element (and all it's children) to the frontend. There it gets updated accordingly. And interactions like klick and select send the information to the backend so a Python function can be called to do what needs to be done.

**Reply by yipeng-jiang** [reply]
Thank you very much. It is amazing.  I just implemented the .update() method into my app. The loading speed is extremely fast like everything is running on the client side. It is definitely much faster than streamlit to rerun everything.  Also, thank you very much for the detailed doc strings on each methods, it helps me a lot.
